design_notes/v2.md

```diff
@@ -0,0 +1,173 @@
+- Modular code base that uses Obspec for the object storage interface abstraction, Zarr as the core data model, and Xarray for the concatenation and merging of Zarr stores
+
+## Related issues
+
+- [Fragility of url auto-parsing logic](https://github.com/zarr-developers/VirtualiZarr/issues/561)
+- [Function definition for open_virtual_dataset in next major release](https://github.com/zarr-developers/VirtualiZarr/issues/553)
+- [Support passing configuration options to default_object_store](https://github.com/zarr-developers/VirtualiZarr/issues/559)
+- [Redefine Virtual Readers as func(filepath) -> ManifestStore](https://github.com/zarr-developers/VirtualiZarr/issues/498)
+- [Make xarray an optional dependency?](https://github.com/zarr-developers/VirtualiZarr/issues/521)
+- [Re-implement loadable_variables using ManifestStore ](https://github.com/zarr-developers/VirtualiZarr/issues/473)
+- [Should we try to use Obstore for reading files?](https://github.com/zarr-developers/VirtualiZarr/issues/476)
+- [Challenges with earthaccess.open_virtual_mfdataset for grouped netCDFs with nested structure](https://github.com/zarr-developers/VirtualiZarr/issues/487)
+- [Docs on how to write a custom reader](https://github.com/zarr-developers/VirtualiZarr/issues/452)
+- [Make readers pluggable via entrypoint system](https://github.com/zarr-developers/VirtualiZarr/issues/245)
+- [Make Manifests' in-memory reference structure pluggable](https://github.com/zarr-developers/VirtualiZarr/issues/246)
+- [Architecture diagram](https://github.com/zarr-developers/VirtualiZarr/issues/225)
+
+## Data structures
+
+`ChunkManifest` - Contains paths, chunk byte offsets, lengths
+`ManifestArray` - Contains one ChunkManifest and Zarr V3 ArrayMetadata
+`ManifestGroup` - Contains group metadata and ManifestGroup and/or ManifestArrays
+`ManifestStore` - Contains a ManifestGroup and a FileReader
+`xarray.Dataset` - Contains ManifestArray-backed variables
+
+## Protocols
+
+> **Note**: Protocols over ABCs would allow third party backends to be less tightly coupled to VirtualiZarr
+> **Question:**: How do we simplify these protocol definitions using obspec?
+
+`FileReader` - read bytes from files on object storage or disk
+`VirtualZarrBackend` - construct a ManifestStore from a file
+
+```python!=
+class FileReader(protocol):
+    """
+    Protocol for file operations for downstream libraries
+    """
+
+    def read(self, size: int, /) -> bytes
+        ...
+
+    def seek(self, offset: int, whence: int = 0, /):
+        ...
+
+    def tell(self) -> int:
+        ...
+
+    async get_range_async(
+        path: str,
+        *,
+        start: int,
+        end: int | None = None,
+        length: int | None = None
+    ) -> Bytes :
+        ...
+
+    async get(
+        path: str,
+        *,
+        options: int | None = None,
+        length: int | None = None
+    ) -> GetResult :
+        ...
+
+```
+
+
+```python!=
+class VirtualZarrBackend(protocol):
```

Comment:
Not everything will be a protocol. Only the "common interfaces" should be defined as protocols. 

E.g. if you only ever have one concrete implementation of "opening a virtual store", then `open_virtual_store` can be a top-level function, not its own protocol.

Here, `open_virtual_store` might _use_ the above FileReader. So you'd instead have something like `open_virtual_store(path: str, *, reader: FileReader)`.

Reply:
A crucial concept here is that the end user needs to be able to pass in the protocol as a parameter in the top-level API somehow. The idea of the protocol is to make it easier to be _externally-implementable_. If you have a "closed" selection of types, where you know the only supported backends will be ones implemented in virtualizarr, then you don't necessarily need to go through the protocol definitions.

It's maybe ok if there's some default behavior that constructs a default filereader, but it needs to be overridden.

But right now you define `FileReader` as a protocol but never define it as a parameter in this document. What uses the `FileReader`? `open_virtual_store`?

Reply:
@kylebarron for `VirtualZarrBackend` at least the motivating context is here: https://github.com/zarr-developers/VirtualiZarr/issues/498. I agree though that it doesn't need to be very complicated.